tests/conftest.py

```diff
@@ -2,29 +2,254 @@
 
 from __future__ import annotations
 
-from typing import AsyncIterator
+import asyncio
+from copy import copy
+from pathlib import Path
+from typing import AsyncIterator, Iterator
 
+import pytest
 import pytest_asyncio
+import structlog
 from asgi_lifespan import LifespanManager
 from fastapi import FastAPI
 from httpx import AsyncClient
+from safir import logging
+from safir.kubernetes import initialize_kubernetes
+from structlog.stdlib import BoundLogger
 
-from jupyterlabcontroller import main
+from jupyterlabcontroller.config import Configuration
+from jupyterlabcontroller.dependencies.config import configuration_dependency
+from jupyterlabcontroller.dependencies.prepull import (
+    prepuller_manager_dependency,
+)
+from jupyterlabcontroller.dependencies.storage import (
+    docker_storage_dependency,
+    gafaelfawr_storage_dependency,
+    k8s_storage_dependency,
+)
+from jupyterlabcontroller.main import create_app
+from jupyterlabcontroller.models.context import Context
+from jupyterlabcontroller.models.domain.docker import DockerCredentialsMap
+from jupyterlabcontroller.models.domain.storage import StorageClientBundle
+from jupyterlabcontroller.models.domain.usermap import UserMap
+from jupyterlabcontroller.models.v1.lab import UserInfo
+from jupyterlabcontroller.services.prepuller import PrepullerManager
+from jupyterlabcontroller.storage.docker import DockerStorageClient
+from jupyterlabcontroller.storage.gafaelfawr import GafaelfawrStorageClient
+from jupyterlabcontroller.storage.k8s import K8sStorageClient
+
+from .settings import TestObjectFactory, test_object_factory
+from .support.mockdocker import MockDockerStorageClient
+from .support.mockgafaelfawr import MockGafaelfawrStorageClient
+from .support.mockk8s import MockK8sStorageClient
+
+_here = Path(__file__).parent
+
+CONFIG_DIR = Path(_here / "configs/standard")
```

Comment:
You will probably eventually need to pick from a set of different configuration files, since you'll want to test GAR vs. Docker Hub, cases with and without an init container, etc. You can wait to add that machinery until you need it, but wanted to flag it here as something to start thinking about.

Reply:
I'm not really sure how I want to do that.  Ideally I'd generate a new set of fixtures for each config dir--the ability to have multiple dirs is why it's configs/standard rather than just config.  But I'm not convinced I want to re-run each test for each configuration.